### PR TITLE
(resolve #1) gk_prototype emberjs

### LINKED ISSUE #1: emberjs prototype
This issue is related to [project](https://github.com/geokrety/geokrety-frontend/projects/1) task: "create emberjs prototype".

A PR must be done to deliver a prototype. 

Prototype acceptance criteria are listed in related project.
This issue would help to collect feedback about criteria or prototype itself.

### PR DESCRIPTION
**Pull request by boly38**
(resolve #1) gk_prototype emberjs, use online api's mock (apiary.io), docker+apache proxy, devops shells, tests sample

### CODE REVIEW

**Comment by boly38** [comment]
@kumy is it possible for you to enable travis-ci for this repo ? thx

File: env/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,37 @@
+# doc : https://docs.docker.com/engine/reference/builder/#shell
+# start from image
+FROM webdevops/apache:latest
```
[the remark this answers is not in the document]
Comment (kumy) [reply env/Dockerfile]: ok, seen code later, you are proxying `https://private-1b51bc-geokretsapiv2mocks.apiary-mock.com/v1/`, however you don't seem to use it `https://github.com/geokrety/geokrety-frontend/pull/2/files#diff-0b8947d92a6233d134fe5979954d7ec6R35`

File: env/exposed/docker/apache2.conf [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+# "LogLevel info ssl:warn"
+#
+LogLevel warn
+
+# Include module configuration:
+IncludeOptional mods-enabled/*.load
+IncludeOptional mods-enabled/*.conf
+
+# Include list of ports to listen on
+Include ports.conf
```
Comment (kumy) [inline env/exposed/docker/apache2.conf]: Bad EOL

File: env/exposed/docker/apache2.conf [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+LogFormat "%{User-agent}i" agent
+
+# Include of directories ignores editors' and dpkg's backup files,
+# see README.Debian for details.
+
+# Include generic snippets of statements
+IncludeOptional conf-enabled/*.conf
+
+# Include the virtual host configurations:
+IncludeOptional sites-enabled/*.conf
```
Comment (kumy) [inline env/exposed/docker/apache2.conf]: Bad EOL

File: env/launch.sh [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+PROTODIR=/app
+echo " * docker-machine create $MACHINE:$PROTODIR"
+docker-machine ssh $MACHINE "sudo mkdir -p $PROTODIR && sudo chown docker $PROTODIR && sudo chmod 777 $PROTODIR" || die "Unable to create $MACHINE $PROTODIR"
+docker-machine scp -r exposed/www/* $MACHINE:/app 2>/dev/null 1>/dev/null || alternate_scp
+
+echo " * Docker compose ($COMPOSE_FILE)"
+docker-compose -f $COMPOSE_FILE up -d --force-recreate || die "compose error"
+
+echo "serving ./exposed/ as https://${GKFRONT_IP}/"
+explorer "https://${GKFRONT_IP}/" || echo "done" # avoid exit 1 status with explorer
```
Comment (kumy) [inline env/launch.sh]: My heart stopped one second ;) 💔

File: gk_prototype/README.md [hunk 5]
```diff
@@ -0,0 +1,64 @@
+# GKFrontend prototype(s)
+
+
+## HowTo run prototype ?
+
+### setup dependencies
+
+	(prototype dir)/DEV_setup.sh
```
Comment (kumy) [inline gk_prototype/README.md]: Space or tab?

File: gk_prototype/README.md [hunk 6]
```diff
@@ -0,0 +1,64 @@
+# GKFrontend prototype(s)
```
Comment (kumy) [inline gk_prototype/README.md]: Nice idea this file :+1:

File: gk_prototype/emberjs/github-fastboot-example/.editorconfig [hunk 7]
```diff
@@ -0,0 +1,20 @@
+# EditorConfig helps developers define and maintain consistent
+# coding styles between different editors and IDEs
+# editorconfig.org
+
+root = true
+
+
+[*]
```
Comment (kumy) [inline gk_prototype/emberjs/github-fastboot-example/.editorconfig]: Will those coding style definitions be checked in Unit tests?

File: gk_prototype/emberjs/github-fastboot-example/.travis.yml [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+  global:
+    # See https://git.io/vdao3 for details.
+    - JOBS=1
+
+before_install:
+  - npm config set spin false
+
+script:
+  - npm run lint:js
+  - npm test
```
Comment (kumy) [inline gk_prototype/emberjs/github-fastboot-example/.travis.yml]: Travis is now enabled for this repo, next push should launch the tests

File: gk_prototype/emberjs/github-fastboot-example/app/routes/geokrets.js [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+    // requirement: ember serve --proxy https://private-1b51bc-geokretsapiv2mocks.apiary-mock.com
+    var urlInfo = "https://geokretsapiv2mocks.docs.apiary.io/ :: /v1/geokrety-type";
+    Logger.info('fetck geokrets list from ' + urlInfo);
+    var gklist = fetch('/v1/geokrety-type')
+                        .then(function(response) {
+                                // Logger.info('response', response);
+                                return response;
+                             })
+                        ;
```
Comment (kumy) [inline gk_prototype/emberjs/github-fastboot-example/app/routes/geokrets.js]: I'm a bit surprised about this part, why not use `ember-data`, coupled with `JSONAPI` it is extremely powerful

`app/routes/news-details.js`
```
import Ember from 'ember';
import News from '../models/news';

export default Ember.Route.extend({
  model(params) {
    return this.get('store').findRecord('news', params.news_id, {
      // include: 'comments,comments.author,author',
      include: 'comments,comments.author,author',
      limit: 1,
      order: '-created-on-date-time'
    });
  }
});
```

`app/models/news.js`

```
import DS from 'ember-data';
import {
  validator,
  buildValidations
} from 'ember-cp-validations';

const Validations = buildValidations({
  comments: validator('belongs-to'),
  content: {
    description: 'content',
    validators: [
      validator('presence', true),
      validator('length', {
        min: 10,
        max: 125
      })
    ]
  },
});

export default DS.Model.extend(Validations, {
	date: DS.attr('string'),
	// czasPostu: DS.attr('string'),
	date: DS.attr('string'),
	title: DS.attr('string'),
	content: DS.attr('string'),
	username: DS.attr('string'),
	userId: DS.attr('number'),
	commentsCount: DS.attr('number'),
	lastComment: DS.attr('string'),
	author: DS.belongsTo('user', {
    inverse: 'news',
    async: false
  }),
	comments: DS.hasMany('newsComment', {
    inverse: 'news',
    async: false
  }),
});
```


In template `{{news-show news=model}}` and the component

```
<div class="panel panel-default">
  <div class="panel-heading">
    <h3 class="panel-title">{{news.title}}</h3>
  </div>
  <div class="panel-body">
    {{{news.content}}}
  </div>
  <div class="panel-footer">
    <div class="pull-right">
      {{#link-to "news-details" news.id}}Comments ({{news.commentsCount}}){{/link-to}}
      -
      {{moment-format news.date 'MMMM D, YYYY'}}
      -
      {{#link-to-user user=news.author}}{{news.username}}{{/link-to-user}}
    </div>
    <div class="clearfix"></div>
  </div>
</div>
```

I'll publish my test with ember after the review

File: gk_prototype/emberjs/github-fastboot-example/app/routes/gitlabuser.js [hunk 10]
```diff
@@ -0,0 +1,17 @@
+import Ember from 'ember';
+import Route from '@ember/routing/route';
+import fetch from 'ember-fetch/ajax';
+
+const {Logger}= Ember;
+
+export default Route.extend({
+	model() {
+		var githubuser = 'kumy'
```
Comment (kumy) [inline gk_prototype/emberjs/github-fastboot-example/app/routes/gitlabuser.js]: haha!

File: gk_prototype/emberjs/github-fastboot-example/app/styles/app.css [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+}
+
+.gk-list h1 {
+	font-size: 12pt;
+	background-color: lightyellow;
+}
+
+
+.kretfoto {
+    background: url(https://raw.githubusercontent.com/geokrety/GeoKrety-Static/master/content/images/help/kret-foto.jpg);
```
Comment (kumy) [inline gk_prototype/emberjs/github-fastboot-example/app/styles/app.css]: Don't hesitate to use our "cdn" https://cdn.geokrety.org/images/help/kret-foto.jpg

File: gk_prototype/emberjs/github-fastboot-example/mirage/config.js [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+
+    this.get('/posts');
+    this.post('/posts');
+    this.get('/posts/:id');
+    this.put('/posts/:id'); // or this.patch
+    this.del('/posts/:id');
+
+    http://www.ember-cli-mirage.com/docs/v0.3.x/shorthands/
+  */
+  this.get('/api/users', function() {
```
Comment (kumy) [inline gk_prototype/emberjs/github-fastboot-example/mirage/config.js]: Those data are not related to our app, but mirage is really nice in mocking in emberJs context!

Check: https://github.com/kumy/gk-front-ember-test/blob/master/mirage/config.js

**Comment by kumy** [comment]
@boly38 `travis-ci` enabled. https://travis-ci.org/geokrety/geokrety-frontend

**Comment by boly38** [comment]
thanks for your amazing feedbacks @kumy :)
will work on it asap
I was not sure about PR. Sounds You're right: will move that on another repository. And so look at your test

File: gk_prototype/emberjs/github-fastboot-example/app/templates/geokrets.hbs [hunk 13]
```diff
@@ -0,0 +1,9 @@
+<!-- possible todo: use controller, view! -->
+<div class="jumbo">
+  <div class="right"></div>
+  <h2>{{model.meta.count}} geokret{{#if (compare model.meta.count '>' 1)}}s{{/if}}</h2>
```
[the remark this answers is not in the document]
Comment (boly38) [reply gk_prototype/emberjs/github-fastboot-example/app/templates/geokrets.hbs]: {{pluralize 2 "geokret"}} gives "geokrets" :P
{{pluralize 2 "geokret" ":pl"}} is not implemented

File: env/Dockerfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (boly38) [reply env/Dockerfile]: not sure to understand your "however" point
* using ember serve --proxy, when geokrety's route will make a fetch, ember will proxy the query
* using apache container, when geokrety's route will make a fetch, apache http proxy will proxy the /v1/* query

File: gk_prototype/emberjs/github-fastboot-example/.editorconfig (hunk 7 above)
Comment (boly38) [reply gk_prototype/emberjs/github-fastboot-example/.editorconfig]: no idea

File: gk_prototype/emberjs/github-fastboot-example/app/routes/geokrets.js (hunk 9 above)
Comment (boly38) [reply gk_prototype/emberjs/github-fastboot-example/app/routes/geokrets.js]: thanks for the sample, I have to look at this feature ! yeah

File: gk_prototype/emberjs/github-fastboot-example/app/templates/geokrets.hbs (hunk 13 above)
[the remark this answers is not in the document]
Comment (boly38) [reply gk_prototype/emberjs/github-fastboot-example/app/templates/geokrets.hbs]: Inflector: just in addition, to make it working from template I had to importe rules from `app.js` (as in [off. doc](https://guides.emberjs.com/v2.11.0/models/customizing-adapters/#toc_pluralization-customization) )

    // sets up Ember.Inflector
    import './models/geokrety-inflector-rules';

**Comment by boly38** [comment]
I had create dedicated repo to avoid merging this PR
https://github.com/boly38/geokrety_emberjs_prototype

**Pull request closed without merging**